File: loan_simulator/account.py


def compute_monthly_interest(amount, interest_rate, offset_amount=0):
    """
    The interest charged for a given month, at the current amount.
    """
    # TODO: Take care of the offset amount causing the balance to become zero. Handle negative amounts.
    monthly_interest_rate = interest_rate / 12
    monthly_interest = (amount - offset_amount) * monthly_interest_rate
    return monthly_interest

class Account:
    def __init__(self, initial_balance):
        self.balance = initial_balance
        self.balance_history = [initial_balance]

    def deposit(self, amount):
        self.balance += amount

    def withdraw(self, amount):
        if self.balance >= amount:
            # The funds are available.
            self.balance -= amount
            return amount
        else:
            # Not enough funds are available.
            amount = self.balance
            self.balance = 0
            return amount

    def get_balance(self):
        return self.balance

    def process_transactions(self):
        # Perform any monthly transactions required of this account.
        pass

    def accrue_interest(self):
        # Generate interest.
        pass

    def get_balance_history(self):
        return self.balance_history

File: wordpress/loan-simulator/src/account.js

export default class Account {
    constructor(start_date, initial_balance){
        this.balance = initial_balance;
        this.balance_history = [];
        // TODO: Remove this negative. It's only there temporarily to display the graph.
        //this.balance_history.push([start_date, initial_balance]);
    }

    deposit(amount) {
        this.balance += amount
    }

    // Move to Account base class.
    withdraw(amount) {
        if (this.balance >= amount) {
            // The funds are available.
            this.balance -= amount;
            return amount;
        } else {
            // Not enough funds are available.
            amount = this.balance;
            this.balance = 0;
            return amount;
        }
    }

    get_balance() {
        return this.balance
    }

    process_transactions(current_date) {
        // Perform any monthly transactions required of this account.
        //pass
    }

    accrue_interest() {
        // Generate interest.
        //pass
    }

    get_balance_history() {
        return this.balance_history
    }

}